csrc/scheduler/normalization_inner.cpp

```diff
@@ -252,6 +263,335 @@ int64_t getMaxPersistentBatch(
+      ceilDiv(reduction_count_after_vectorize, persistent_batch_size) *
+          persistent_batch_size -
+      reduction_count_after_vectorize;
+  return params;
+}
+
+// Return true if ha is better than hb
+
+// This sorting function ensures the selected heuristic meeting target
+// occupancy, promotes even workload distribution, enhances register
+// optimization, and prefers higher occupancy.
+// TODO: It leads to 10% regression for softmax around 2K to 6K and 16K.
+// See https://github.com/NVIDIA/Fuser/issues/1876
+bool compareTwoHeuristics(
+    const HeuristicParams& ha,
+    const HeuristicParams& hb,
+    const int64_t target_register_overhead,
+    const int64_t target_warps_per_sm) {
+  auto compare = [](int64_t a, int64_t b) -> int {
+    return a > b ? 1 : (a < b ? -1 : 0);
+  };
+  int score = 0;
+
+  // prefer occupancy larger than target
+  score = compare(
+      ha.occupancy >= target_warps_per_sm, hb.occupancy >= target_warps_per_sm);
+  if (score != 0) {
+    return score > 0;
+  }
+
+  // prefer reduction count after vectorization is divisible by persistent
+  // batch size
+  score = compare(ha.n_persistent_tails == 0, hb.n_persistent_tails == 0);
+  if (score != 0) {
+    return score > 0;
+  }
+
+  // Large register overhead leaves more registers for the compiler to
+  // optimize generated SASS code and avoids register spills. But don't want
+  // to use a very large block size which increaes inter-thread communication.
+  // This condition influence dropout_layer_norm at 7K and softmax_dropout at
+  // 6K and 16K.
+  constexpr int64_t opt_max_threads_per_block = 512l;
+  score = compare(
+      ha.free_registers > target_register_overhead &&
+          ha.padded_bdimx <= opt_max_threads_per_block,
+      hb.free_registers > target_register_overhead &&
+          hb.padded_bdimx <= opt_max_threads_per_block);
+  if (score != 0) {
+    return score > 0;
+  }
+
+  // Prefer large occupancy
+  score = compare(ha.occupancy, hb.occupancy);
+  if (score != 0) {
+    return score > 0;
+  }
+
+  // Tiebreaker, use large persistent batch size so more registers are used
+  // for the persistent buffer.
+  return ha.persistent_batch_size > hb.persistent_batch_size;
+}
+
+// Generate a heuristic for each possible persistent batch size.
+// (1) If the maximum occupancy is less than the target occupancy, use the batch
+//     leads to the largest occupancy.
+// (2) sort the heuristics as follows:
+//     (a) Prioritize occupancy exceeding target.
+//         Ensures minimum required occupancy is surpassed.
+//     (b) Favor divisibility by persistent batch size.
+//         Aims for even workload distribution.
+//     (c) Opt for more free registers.
+//         Maximizes compiler optimization potential.
+//     (d) Seek larger occupancy.
+//         Exceeds the target minimum for better performance.
+//     (e) Use large persistent batch size as a tiebreaker.
+//         Use more registers for persistent buffers.
```

Comment:
One last word here is that, I don't think we have gone through great length in discussing the criterion here. But our end performance looked pretty good so I'm fine with us moving forward with this. :stuck_out_tongue_winking_eye: